### ISSUE TITLE
Should token.includes_scope? return false if not all required scopes are included?

### ISSUE THREAD

**Issue by calfzhou** [doorkeeper-gem/doorkeeper#591]
```ruby
      def includes_scope?(*required_scopes)
        required_scopes.blank? || required_scopes.any? { |s| scopes.exists?(s.to_s) }
      end
```
By the above code, `includes_scope?` will success even if one or more required scopes are not included. I think this behavior is ambiguous, since scopes is a set, we say a set A includes another one B if and only if every item in set B exists in set A, e.g. { x, y, z } includes { x, y }, but *not* includes { x, w }.

Thus, if one action is protected by 
```ruby
doorkeeper_authorize! :read, :write
```

I think only when the client's authorized scopes include both `:read` and `:write`, it can access this action.

Basically, there should be two checkers, `includes_all_scopes?` and `includes_any_scope?`, and by default it should make sure the client's scopes include *all* required ones.

**Comment by calfzhou** [comment]
I just checked swagger spec, and i believe that there should be a logical AND between required scopes for each endpoint.

And in OAuth 2.0 rfc doc, [accessing protected resources](https://tools.ietf.org/html/rfc6749#section-7), it saids

> The resource server MUST validate the access token and ensure that it has not expired and that **its scope covers the requested resource**.

which should mean that the granted scope must be greater than or equal to the scope required by the resource.

So the default behavior of
```ruby
doorkeeper_authorize! :read, :write
```
should be ensure that the client have both `:read` and `:write` document.

**Comment by tute** [comment]
@calfzhou thank you for your input. I have other issues to respond to you, but the past and upcoming week look busy for me, so it takes a few days.

`AccessTokenMixin#scopes_match` calls `ScopeChecker#match?`, which uses `has_scopes?`, which has for implementation `scopes.all? { |s| self.exists?(s) }`, which makes sense for authorizing.

Regarding `AccessTokenMixin#includes_scope?` you point out, it comes all the way from 2011 (https://github.com/doorkeeper-gem/doorkeeper/commit/eac1edfd). I just added specs specifying it's "OR" nature: https://github.com/doorkeeper-gem/doorkeeper/commit/9ebd8b31fa51842a60f969b90f3d8d7008b148df.

I agree it is surprising: the API doesn't tell programmers if it will check for any or for all. That is a design mistake; this should be obvious to programmers. Moreover, if a programmer wants an AND operation today, they'd need to define a new `read_write` scope.

I'm not sure, though, that we want to change the semantics of that call. doorkeeper behaves that way since it exists. If anything, I'd introduce syntax to make it obvious, something like `doorkeeper_authorize! all: %w(read write)` (or nicer), but then it would be backwards incompatible.

Short term, then, I would add an explanation on the wiki page and the README about this behavior, and keep it as is. What do you think?

Thank you again for your input.

**Comment by tute** [comment]
ping @calfzhou

**Comment by calfzhou** [comment]
It's ok to just clarify in document that there is a logical **OR** between scopes for authorizing, and `includes_scope?` will success if **at least** one scope requirement meets.

I'm still trying to find out what's the *standard behavior* for checking scopes. But for now, i'm all right to use logical OR for scope checking.

For next step, there could be a new feature allowing user to choose AND or OR freely, without breaking existing behavior.

**Comment by tute** [comment]
Sounds good. Would you like to send a PR to the README please?

**Comment by calfzhou** [comment]
My pleasure. Maybe several days later, i'm recently busy on working :sweat_smile:

**Comment by tute** [comment]
Thanks for the PR!
